README
INSTRUCTIONS
------------

- create the architecture to store the various contributions. From the proceeding root directory the architecture  must follow proceeding_ROOT/contrib/SXX/Author_SXX/

- Move the contributions (the tex and bib files and the associated PDF figures) into the sub-directory contrib/SXX/Author_SXX/ where SXX is the session of the contribution (S01, S02, ...,S20...) and run the script check_contrib.py on the files. It will make a copy of the file, translate the contribution to pure latex, verify the existence of the figures and test compile the contribution.


- Edit the file book-setup.py :
	- update the year
	- update the clsfile if needed
	- update the list of editors
	- update the definition of the sessions and contributions
	- fill in the title of the sessions

- Link the foreword tex file (written by the president of the SF2A) or just rename it
	ln -s your_foreword_file.tex input_foreword.tex
Or
	mv your_foreword_file.tex input_foreword.tex

- Edit book_cover.tex (the file handling the cover page)
	- update the list of editors
	- insert the conference picture/logo etc
	- update location and dates
	- add sponsors logo

- Create the SXX.tex files (head of chapter) in contrib/SXX/ (XX is the session number). To do this run the script make_sessions.py (warning python 3.6+)

- Compile all contributions:
  The compilation is performed by the script make-contrib.py
  The LaTex files are compiled using pdflatex. The bibliography of a given contribution is generated by running bibtex (the scripts checks for the existence of a .bib file to run bibtex).    
  In case of errors during the compilation, correct appropriately the contribution that causes the error and run manually the sequence pdflatex yourfile; bibtex yourfile; pdflatex yourfile; pdflatex yourfile
  If the error was caused by large figures, this problem will appear again while compiling the full proceeding as the inclusion of the pdf will cause a timeout and this message will appear: 'xdvipdfmx:fatal: pdf_link_obj(): passed invalid object.' but not while processing the file in question. To solve this issue, rewrite the pdf yourfile.pdf using gs. Follow this steps:
  cp yourfile.pdf yourfile.bak.pdf
  gs -o yourfile.pdf -sDEVICE=pdfwrite -dPDFSETTINGS=/prepress yourfile.bak.pdf
  

- Book setup:
	Run the Python script make-book.py. This script build the list of contributions (abstracts_ads.txt and index.html) and the file inputpdf.tex (see outputs). 
	Once executed, look at the content of make-book.log to check the way the contributions were included.
	The script may have changed the page counter of one or several contributions. In that case the script will ask the permission to execute the command 'make-contrib.py'

- Edit the inputdpdf.tex file (check for misspelling and bad indexes)

- Edit appropriately the files book_packages.tex, liste_participants.tex and book.tex

- Generate the book (book.pdf) and associated files (contents.pdf, participants.pdf, authors_index.pdf) by typing:
	touch book.ind
	xelatex book
	makeindex book
	xelatex book
	xelatex book

- Create the archive for the website:
	cd contrib
	tar rhvf book.tar book.pdf participants.pdf contents.pdf authors_index.pdf foreword.pdf index.html 
	gzip -f book.tar

- Once the proceedings are online, send abstract_ads.tex to ADS. Follow the procedure describe at this location: https://ui.adsabs.harvard.edu/help/data_faq/conf_proc

OUTPUTS
-------

- the directory contrib/ will  contain all the individual contribution as a link to the contribution in subfolders in PDF format.

- book.pdf : the whole book (including the TOC, list of participants, the contributions and the authors index)

- contents.pdf : the table of contents (TOC)

- participants.pdf : the list of participants

- authors_index.pdf : the author index

- abstracts_ads.txt : the list of references for the SAO/NASA ADS (this list includes the URLs pointing toward the individual contributions)

- inputpdf.tex : This file is generated by make-book.py and contains the links to the contributions in PDF format. A link to this file is included in book.tex

- liste_contrib.html : the list of contributions in HTML format - used to create the web page on the SF2A website.


TOOLS
-----

check_contrib.py : rewrites the tex file as a pure tex file and runs various checks

make-sessions.py : creates the sessions cover pages

make-contrib.py : processes all the contributions

make-book.py : prepares the book setup

make-listparticipants.py : generates the list of participants


RELEASES.md
# SF2Aoroc releases

## v0.1.0-beta, first pre-release
* 2023-09-11 * 


book-setup.py
#
# book-setup.py
# 
# We setup here the book structure and
# define the variables used by the Python script make-book.py  
# 
# Edit this file appropriately
#
# updated by R. Samadi 22.10.2010
#

year ='2022'

bstfile = 'aa.bst'
clsfile = 'sf2a-conf2022'

# the place where the proceedings will be installed
#urlbase_proc="http://lesia.obspm.fr/semaine-sf2a/2012/proceedings/"
#urlbase_proc="http://proc.sf2a.asso.fr/"
#urlbase_proc="http://sf2a.cesr.fr/"
urlbase_proc="http://proceedings.sf2a.eu/" + year + "/"

editeurs='J. Richard, A. Siebert, E. Lagadec, N. Lagarde, O. Venot, J. Malzac, J.-B. Marquette, M. N\'Diaye, D. Briot'

start_page =1

NOCONTRIB=0 # 1 -> the contributions are not inserted, only the chapters are processed

# definition of the 2022 sessions and contribution
# sort out contributions according to Review, Talk, Posters
# (alphabetical order within each subclass)
# comment out sessions with no contribution
S0 = ['S0','Chauvin_S0','Douspis_S0','Knodlseder_S0','Meunier_S0','Ouazzani_S0','Robert_S0']
S1 = ['S1','Boughelilba_S1','DehiwalageDon_S1','ElMellah_S1','Freour_S1','Soudais_S1']
S2 = ['S2','Amard_S2','Cointepas_S2','Dhouib_S2','ElMellah_S2','Maimone_S2','Merle_S2','Meunier_S2','Noraz_S2','Tabone_S2','Thomassona_S2','Thomassonb_S2']
S3 = ['S3','LeGouellec_S3','Schirmer_S3','Taillard_S3']
S4 = ['S4','Boisse_S4','Vienne_S4']
S5 = ['S5','Alkan_S5','Desmars_S5','Robert_S5','Thuillot_S5','Turpin_S5','Delisle_S5','LeCam_S5','Lekic_S5','Marquette_S5','Wullaert_S5']
S6 = ['S6','Webb_S6']
S7 = ['S7','Freour_S7']
S9 = ['S9','Pitout_S9']
S10 = ['S10','Galliano_S10','Seillé_S10']
S11 = ['S11','Breton_S11','Delorme_S11']
S13 = ['S13','Berteaud_S13','Bourgoin_S13','Bruel_S13','Bugli_S13','Contini_S13','Godet_S13','Mignon-Risse_S13','Srinivasan_S13','Webb_S13']
S14 = ['S14','Tresse_S14']
S15 = ['S15','Lemasle_S15']
S16 = ['S16','Dyrek_S16','Malin_S16','Acuna_S16','Desgrange_S16','Ducreux_S16','Jaziri_S16']
S18 = ['S18','Beust_S18','Griveaud_S18','Gry_S18','Benseguane_S18','Cabral_S18','Crida_S18','Gkotsinas_S18','Martel_S18','Michoulier_S18']
S20 = ['S20','Cristofari_S20','Delcamp_S20']
S21 = ['S21','Griessmeier_S21','Cecconi_S21','Bonne_S21','Brionne_S21']
#WARNING : encoding of the ' -> \' and \'e -> \\\'e to keep latex text in the keys
# comment out sessions with no contribution
sessions = [
    {'SF2A --- Plenary session (S0)':S0},
    {'Atelier g\\\'en\\\'eral du PNHE (S1)':S1},
    {'Atelier g\\\'en\\\'eral du PNPS (S2)':S2},
    {'Atelier g\\\'en\\\'eral du PCMI (S3)':S3},
    {'Education \`a l\'astronomie : structuration, activit\\\'es et ressources (S4)': S4},
    {'Collaboration Pro-Am Gemini (S5)': S5},
    {'L\'\\\'egalit\\\'e des genres en A\&A : point sur les jeunes (S6)': S6},
    {'Transition environnementale : quel r\\^ole pour la communaut\\\'e astronomique ? (S7)': S7},
    {'Mesures in-situ et en t\\\'el\\\'ed\\\'etection des plasmas du Syst\\`eme solaire (S9)': S9},
    {'Dust in galaxies: from the local interstellar medium to distant galaxies (S10)': S10},
    {'Le Soleil est-il une \\\'etoile de type solaire ? (S11)': S11},
    {'Gravitational-wave astronomy and multi-messenger astrophysics (S13)': S13},
    {'L\'astrophysique fran\\c{c}aise \\`a l\'heure du JWST : opportunit\\\'es et d\\\'efis (S14)':S14},
    {'The Local Group in the Gaia era: from the Galactic halo to the Andromeda galaxy (S15)': S15},
    {'Atmosph\\`eres d\'exoplan\\`etes dans le contexte du JWST (S16)': S16},
    {'Disques de d\\\'ebris, exocom\\`etes et formation plan\\\'etaire (S18)': S18},
    {'Cosmic turbulence (S20)': S20},
    {'Le renouveau de la radio-astronomie et SKA, c\'est maintenant ! (S21)': S21}
    ]



make-sessions.py
#!/usr/bin/python
# -*- coding: latin-1 -*-
# This program creates the .tex files for the individual sessions
# from the file book-setup.py
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Written in python3 by A. Siebert 8.Oct.2021


import string
import re
from sf2a_utils import *
import os
import re
import sys
import codecs
import subprocess
from pathlib import Path

# Check python version, must be at least 3
if sys.version_info[0] < 3:
    raise Exception("Must be using Python 3+, your system is using Python "+str(sys.version_info[0]))


# Read setup file
with open("book-setup.py") as myf:
    code = compile(myf.read(), "book-setup.py", 'exec')
    exec(code)

# Create template for file
mytemplate=string.Template('\\documentclass{$MYCLS}\n\
\\usepackage{graphicx}\n\n\
\\newcommand{\sigle}{Session $SES_NUM}\n\
\\newcommand{\\texte}{$SES_NAME}\n\
\n\
\\begin{document}\n\
\n\
\TitreGlobal{SF2A $YEAR}\n\
\n\
\\title{NOTINTOC} \n\
%\n\
\\author{}\\address{}\n\
\\runningtitle{}\n\
%\n\
\setcounter{page}{1}\n\
% Keep this line, even if the page will be settled afterwards...\n\
\n\
\\vspace*{5cm}\n\
\n\
\centering{\huge \sigle}\n\
\n\
\\vspace*{2cm}\n\
\n\
\centering\\begin{minipage}[h]{.9\linewidth}\n\
\centering{\huge \\texte}\n\
\end{minipage}\n\
\n\
\\newpage\n\
\n\
\\, \n\
\n\
%\centering\includegraphics[width = .9\linewidth]{photo}\n\
\n\
%\\textsl{}\n\
\n\
\\vfill\n\
\n\
%\centering\includegraphics[width = .9\linewidth]{photo}\n\
\n\
\\vfill\n\
\n\
\end{document}\n')

# Loop over all sessions to create tex file
for num in range(len(sessions)):
    session = sessions[num]
    session_name = list(session.keys())[0]

    # la contrib generale de la session est toujours en premier, pas la peine de boucler pour la chercher
    session_num = list(session.values())[0][0]
    path = 'contrib/' + session_num
    filename = 'contrib/' + session_num + '/' + session_num + '.tex'
    print("Creating tex file for ",session_name," in ",filename)


    texfile = open(filename,"w")
    print(filename,clsfile,session_num,year,session_name)
    mytext=mytemplate.substitute(MYCLS=clsfile,SES_NUM=session_num, YEAR=year, SES_NAME=session_name)
    texfile.write(mytext)
    texfile.close()
    #print(mytext)


unicode2latex.py
#!/usr/bin/env python3


## by A. Mennucci

## starting from
## https://tex.stackexchange.com/a/23438/199657
## ( adapted to Python3)

## added support for characters with multiple accents,
## https://unicode.org/reports/tr15/tr15-27.html

##  and many decompositions
## https://www.compart.com/en/unicode/decomposition/

"""
This script will convert unicode to a suitable LaTeX representation.
It will convert accents, e.g. è  → \`{e} .
It will express fonts, e.g.  ⅅ → \symbbit{D} .
It will convert math symbols, e.g.  ∩ → \cap .
"""

import os, sys, copy, io, argparse, unicodedata, collections, subprocess, logging
logger = logging.getLogger(__name__)

verbose = 0

prefer_unicode_math = False

unicode_math_file = '/usr/share/texlive/texmf-dist/tex/latex/unicode-math/unicode-math-table.tex'

try:
    p = subprocess.Popen(['kpsewhich', 'unicode-math-table.tex'], stdout=subprocess.PIPE)
    a = p.stdout.read().strip()
    p.wait()
    if os.path.isfile(a):
        unicode_math_file = a
    else:
        logger.warning('Cannot locate unicode-math-table.tex')
except:
    logger.exception('While running kpsewhich unicode-math-table.tex')

math2latex = {
    0xD7 : '\\times',
    0x221E : '\\infty',
    0xab : '\\guillemotleft' , 0xbb : '\\guillemotright',
}


## see also
## https://ctan.math.washington.edu/tex-archive/fonts/kpfonts-otf/doc/unimath-kpfonts.pdf

if os.path.isfile(unicode_math_file):
    for s in open(unicode_math_file):
        if s.startswith('\\UnicodeMathSymbol'):
            s = s.split('{')
            code=int(s[1].lstrip('"').rstrip('}').rstrip(' '),16)
            latex=s[2].rstrip('}').rstrip(' ')
            if code not in math2latex: 
                math2latex[code] = latex
            elif verbose:
                sys.stderr.write('(Prefer %r for %x to %r )\n' % (math2latex[code], code,latex))



accents = {
    0x0300: '`', 0x0301: "'", 0x0302: '^', 0x0308: '"',
    0x030B: 'H', 0x0303: '~', 0x0327: 'c', 0x0328: 'k',
    0x0304: '=', 0x0331: 'b', 0x0307: '.', 0x0323: 'd',
    0x030A: 'r', 0x0306: 'u', 0x030C: 'v',
}

##https://ctan.mirror.garr.it/mirrors/ctan/macros/unicodetex/latex/unicode-math/unicode-math.pdf
##table 7
fonts = (
    ('SANS-SERIF BOLD ITALIC', '\\symbfsfit'),
    ('SANS-SERIF BOLD',        '\\symbfit'),
    ('SANS-SERIF ITALIC',      '\\symsfit'),
    ('BOLD ITALIC',  '\\symbfit'),
    ## currently broken
    #('BOLD FRAKTUR', '\\symbffrac'),
    ('BOLD SCRIPT',  '\\symbfscr'),
    ('DOUBLE-STRUCK ITALIC',  '\\symbbit'),
    #
    ('DOUBLE-STRUCK',  '\\symbb'),
    ('BLACK-LETTER',   '\\symfrak'),
    ('FRAKTUR',        '\\symfrak'),
    ('SANS-SERIF',     '\\symsf'),
    ('MONOSPACE',      '\\symtt'),
    ('SCRIPT',         '\\symscr'),
    ('BOLD',           '\\symbf'),
    ('ITALIC',         '\\symit'),
    )
    

modifiers = {
    '<super>': '^{%s}' , '<sub>' : '_{%s}',
}


line_count = 1
char_count = 0
input_file = 'cmdline'


def decompose_to_tex(text, output):
    global char_count
    if isinstance(text,str):
        assert(len(text) == 1)
        char = text
    else:
    #if isinstance(text,  collections.abc.Iterable):
        char = next(text)
        char_count += 1
    code = ord(char)
    cat = unicodedata.category(char)
    dec = unicodedata.decomposition(char)
    name = unicodedata.name(char, '')
    #
    if prefer_unicode_math and code in math2latex and code > 0x7f:
        latex = math2latex[code]
        if verbose: sys.stderr.write('%r:%d:%d math %r %x %s\n' % (input_file, line_count, char_count,char,code,latex))
        output.append(latex+' ')
        return
    #
    if (verbose and dec) or (verbose > 1 ):
        sys.stderr.write('%r:%d:%d decomposing %r %x %r %r %r\n' % (input_file, line_count, char_count,char,code,cat,dec,name))
    #
    if cat in ("Mn", "Mc") and code in accents:
        try:
            n = output.pop()
        except IndexError:
            sys.stderr.write('%r:%d:%d accents with no preceding character\n' % (input_file, line))
            n=' '
        output.append( "\\%s{%s}" %(accents[code], n) )
        return
    
    if dec:
        if dec.startswith('<fraction>'):
            dec =  dec.split()[1:]
            dec = list(map(lambda x : chr(int(x,16)), dec))
            if dec[-1] == '/': dec[-1] = ''
            output.append( '{\sfrac{%s}{%s}}' % (dec[0],dec[-1]))
            return
        decsplit = dec.split()
        base, acc = decsplit[:2]
        if base == '<font>':
            mod = '%s'
            N = copy.copy(name)
            for s,l in fonts:
                if s in N:
                    #print((s,N))
                    mod = l + '{' +  mod + '}'
                    j = N.index(s)
                    N = N[:j] + N[j+len(s):]
                    #print((s,N))
            acc = int(acc, 16)
            decompose_to_tex(chr(acc), output)
            n = output.pop()
            output.append( mod % n )
            return
        if base == '<small>':
            acc = int(acc, 16)
            decompose_to_tex(chr(acc), output)
            n = output.pop()
            output.append( '{\scriptsize{%s}}' % n )
            return
        if base == '<compat>':
            # greek characters, `var` version
            acc = int(acc, 16)
            if 0x370 <= acc <= 0x3ff:
                decompose_to_tex(chr(acc), output)
                n = output.pop()
                output.append( n[0] + 'var' + n[1:] )
                return
            elif 'LIGATURE' in name:
                for l in decsplit[1:]:
                    l = int(l, 16)
                    decompose_to_tex(chr(l), output)
                return
            sys.stderr.write('%r:%d:%d unsupported <compat> %r\n' % (input_file, line_count, char_count,dec))
        elif base in modifiers:
            mod = modifiers[base]
            acc = int(acc, 16)
            decompose_to_tex(chr(acc), output)
            n = output.pop()
            output.append( mod % n )
            return
        elif base.startswith('<'):
            sys.stderr.write('%r:%d:%d unsupported modifier %r\n' % (input_file, line_count, char_count,base))
            #acc = int(acc, 16)
            #decompose_to_tex(chr(acc), output)
            #n = output.pop()
            #output.append( mod % n )
            output.append(char)
            return
        try:
            base = int(base, 16)
            acc = int(acc, 16)
            if acc in accents:
                decompose_to_tex(chr(base), output)
                n = output.pop()
                output.append( "\\%s{%s}" %(accents[acc], n) )
                return
        except ValueError:
            sys.stderr.write('%r:%d:%d unsupported decomposition %r\n' % (input_file, line_count, char_count,dec))
    #
    if name.startswith('GREEK'):
        char = name.split()[-1].lower()
        if 'SMALL' not in name:
            char = char[0].upper() + char[1:]
        output.append('\\' + char)
        return
    #
    if code in math2latex and code > 0x7f:
        latex = math2latex[code]
        if verbose: sys.stderr.write('%r:%d:%d math %r %x %s\n' % (input_file, line_count, char_count,char,code,latex))
        output.append(latex+' ')
        return
    if code > 127:
        sys.stderr.write('%r:%d:%d could not convert %r %x to ascii\n' % (input_file, line_count, char_count,char,code))
    output.append(char)
        
def uni2tex(text):
    iterable = iter(text)
    output = []
    try:
        while True:
            decompose_to_tex(iterable, output)
    except StopIteration:
        pass
    return ''.join(output)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='convert Unicode to LaTeX',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                                     epilog=__doc__)
    parser.add_argument('--verbose','-v',action='count',default=verbose)
    parser.add_argument('--prefer-unicode-math','-P',action='store_true',
                        default=prefer_unicode_math,
                        help='converto to unicode-math commands whenever possible')
    parser.add_argument('--output','-o',
                        help='output file')
    parser.add_argument('--input','-i',
                        help='read input file')
    parser.add_argument('--stdin',action='store_true',
                        help='read stdin (instead of input file)')
    parser.add_argument('text', nargs='*',
                        help='the unicode text')
    args = parser.parse_args()
    #
    if 1 != bool(args.text) + bool(args.input) + bool(args.stdin):
        parser.print_help()
        sys.exit(1)
    #
    verbose = args.verbose
    prefer_unicode_math = args.prefer_unicode_math
    #
    if args.input or args.stdin:
        out=sys.stdout
        if args.output:
            out = open(args.output, 'w')
        if args.text:
            logger.warning('Cmdline %r ignored', args.text)
        if args.stdin:
            input_file = 'stdin'
            for L in sys.stdin:
                out.write(uni2tex(L))
                char_count = 0
                line_count += 1
        else:
            input_file = args.input
            for L in open(args.input):
                out.write(uni2tex(L))
                char_count = 0
                line_count += 1
        sys.exit(0)
    #
    for  t in args.text:
        if not isinstance(t,str):
            t = str(t, "utf-8")
        print(uni2tex(t))
    sys.exit(0)
    


make-listparticipants.py
#!/usr/bin/python
# -*- coding: latin-1 -*-
# This program generates the list of participants of the conference
# based on a provided csv file containing first name, last name and email
# address (other fields will not be used)
#
# Usage : make_listparticipants.py csvfile
#
# Output : list_participants.tex
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Written in python3 by A. Siebert 8.Oct.2021

import string
import re
from sf2a_utils import *
import os
import re
import sys
import codecs
import subprocess
from pathlib import Path
import csv

def format_name(mystring):
    """ 
    format_name(str)

    Format the first names to have an upper case letter in first place and lowercase afterwards
    Handles spaces and - separations between names

    Input: str, a string containing first name(s)
    Output: formated string
    """

    mylist = list(mystring)
    mylist[0] = mylist[0].upper()

    ll=len(mylist)
    for i in range(1,ll,1):
        if mylist[i-1] in {' ','-'}:
            mylist[i] = mylist[i].upper()
        else:
            mylist[i] = mylist[i].lower()

    mystring= ''.join(mylist)
    return mystring

#--------------------------

if sys.version_info[0] < 3:
    raise Exception("Must be using Python 3+, your system is using Python "+str(sys.version_info[0]))


if len(sys.argv)<2:
    raise Exception("Usage: make-listparticipants.py yourcsvfile [delimiter=;]\nDefault delimiter is ';'")

mycsvfile = sys.argv[1]

if len(sys.argv)==2:
    mydel = ';'
else:
    mydel=sys.argv[2]

listparticipants = open('liste_participants.tex','w')
t = string.Template('\\textbf{ $LAST $FIRST } (\\email{ $EMAIL })\\\\\n')

listparticipants.write('\\section*{\\Large\\bf  List of participants}\n\n\\vskip 35truemm\n\n')

with open(mycsvfile, newline='') as csvfile:
    reader = csv.DictReader(csvfile, delimiter=mydel)
    seen = []
    for row in reader:
        ff=str(row['First Name'])+'|'+str(row['Last Name'])+'|'+str(row['Email'])
        if not ff in seen:
            seen.append(ff)
            ## TO DO test First Name to add a capital letter for each name
            myline = t.substitute(FIRST=format_name(row['First Name']),LAST = row['Last Name'].upper(), EMAIL = row['Email'])
            listparticipants.write(myline)

listparticipants.close()
csvfile.close()




make-contrib.py
#!/usr/bin/python
# -*- coding: latin-1 -*-
# This program compiles the tex files of the individual contributions
# from the file book-setup.py
# It runs the sequence pdflatex,bibtex,pdflatex,pdflatex
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Written in python3 by A. Siebert 8.Oct.2021

import string
import re
from sf2a_utils import *
import os
import re
import sys
import codecs
import subprocess
from pathlib import Path

# Test for Python version
if sys.version_info[0] < 3:
    raise Exception("Must be using Python 3+, your system is using Python "+str(sys.version_info[0]))

# Open setup file
with open("book-setup.py") as myf:
    code = compile(myf.read(), "book-setup.py", 'exec')
    exec(code)

#  Open file handle for logs
log = open('make-contrib.log','w')

# Start actual work
wd = os.getcwd() # get current directory
mypattern = re.compile(r'^S\d+$')

list_error=[]

for num in range(len(sessions)):
    session = sessions[num]
    session_name = list(session.keys())[0]
    print("----------------------------------------")
    print(session_name)
    print("----------------------------------------")
    log.write("----------------------------------------\n")
    log.write(session_name+'\n')
    log.write("----------------------------------------\n")

    session_num = list(session.values())[0][0]
        
    for nom in list(session.values())[0]:
        match = mypattern.match(nom)
        
        # Composition des noms de fichiers
        if match:
            path = 'contrib/' + session_num
            filebase = 'contrib/' + session_num + '/' + nom
            bibtex = False
        else :
            path = 'contrib/' + session_num + '/' + nom
            filebase = 'contrib/' + session_num + '/' + nom + '/' + nom
            bibtex = True

        print('Processing '+filebase)
        log.write('Processing '+filebase+'\n')
        fullpath  = wd +'/'+path

        # copying bst and cls files to location
        process = subprocess.run(['cp', bstfile, fullpath+'/.'])
        process = subprocess.run(['cp', clsfile+'.cls', fullpath+'/.'])

        # from that point onwards, we are always working in the directory of the tex files
        os.chdir(fullpath)
        log.write('Changing to directory '+fullpath+'\n')
        
        process = subprocess.run(['pdflatex', nom],
                                 stdout=subprocess.PIPE, 
                                 stderr=subprocess.PIPE,
                                 universal_newlines=True)
        if process.stderr:
            print('Error while processing '+filebase+' using pdflatex \nCheck compilation manually:\n')
            print(process.stderr)
            log.write('Error while processing '+filebase+' using pdflatex \nCheck compilation manually:\n')
            log.write(process.stderr+'\n')
            list_error.append(filebase)
            continue
        else:
            print('pdflatex ok')
            log.write('pdflatex ok \n')

        if bibtex:
            bibpath = Path(nom+'.bib')
            if bibpath.is_file():
                process = subprocess.run(['bibtex', nom],
                                         stdout=subprocess.PIPE, 
                                         stderr=subprocess.PIPE,
                                         universal_newlines=True,
                                         cwd=wd+'/'+path)
                if process.stderr:
                    print('Error while processing '+filebase+' using bibtex \nCheck compilation manually:\n')
                    print(process.stderr)
                    log.write('Error while processing '+filebase+' using bibtex \nCheck compilation manually:\n')
                    log.write(process.stderr+'\n')
                    list_error.append(filebase)
                    continue
                else:
                    print('Bibtex ok')
                    log.write('Bibtex ok\n')
            else:
                print('No bibtex file')
                log.write('No bibtex file\n')
                
        process = subprocess.run(['pdflatex', nom],
                                 stdout=subprocess.PIPE, 
                                 stderr=subprocess.PIPE,
                                 universal_newlines=True,
                                 cwd=wd+'/'+path)
        if process.stderr:
            print('Error while processing '+filebase+' using pdflatex \nCheck compilation manually:\n')
            print(process.stderr)
            log.write('Error while processing '+filebase+' using pdflatex n2\nCheck compilation manually:\n')
            log.write(process.stderr+'\n')
            list_error.append(filebase)
            continue
        
        process = subprocess.run(['pdflatex', nom],
                                 stdout=subprocess.PIPE, 
                                 stderr=subprocess.PIPE,
                                 universal_newlines=True,
                                 cwd=wd+'/'+path)
        if process.stderr:
            print('Error while processing '+filebase+' using pdflatex \nCheck compilation manually:\n')
            print(process.stderr)
            log.write('Error while processing '+filebase+' using pdflatex n3\nCheck compilation manually:\n')
            log.write(process.stderr+'\n')
            list_error.append(filebase)
            continue

if len(list_error):
    print("We have encountered some errors while processing the contributions\nPlease check the following directories (more information in the logfile make-contrib.log):\n")
    for contrib in list_error:
        print(contrib+'\n')
else:
    print("Apparently no errors were reported! Good news!\n Check the logfile make-contrib.log anyway...\n")
log.close()
